### PR TITLE
[Security] Add finer control over script checks

### PR DESCRIPTION
**Pull request by Aestek**
Implementation for https://github.com/hashicorp/consul/issues/4711

## Changes

Add `-enable-local-script-checks` option to the agent configuration. This is equivalent to `-enable-script-checks` but only for scripts defined in config files.

## Implementation

`AddService()` and `AddCheck()` now take an extra parameter `source` to signal them whether this call originates from the API or a config read.
This parameter was added to ensure that `AddCheck()` calls nested in the stack (for example `AddService()` calls `AddCheck()`) take the appropriate action based on configuration.
Another reason is to clearly indicate the callers that the func behaves differently depending on this parameter.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -1851,8 +1858,14 @@ func (a *Agent) AddCheck(check *structs.HealthCheck, chkType *structs.CheckType,
 			return fmt.Errorf("Check is not valid: %v", err)
 		}
 
-		if chkType.IsScript() && !a.config.EnableScriptChecks {
-			return fmt.Errorf("Scripts are disabled on this agent; to enable, configure 'enable_script_checks' to true")
+		if chkType.IsScript() {
+			if source == ConfigSourceLocal && !a.config.EnableScriptChecks && !a.config.EnableLocalScriptChecks {
```
Comment (pierresouchay) [inline agent/agent.go]: I am not a fan of having several semantics into the code. I would perform the change to only check for a.config.EnableLocalScriptChecks or EnableRemoteScriptChecks

File: agent/agent.go [hunk 2]
```diff
@@ -1851,8 +1858,14 @@ func (a *Agent) AddCheck(check *structs.HealthCheck, chkType *structs.CheckType,
...
+				return fmt.Errorf("Scripts are disabled on this agent; to enable, configure 'enable_script_checks' or 'enable_local_script_checks' to true")
+			}
+
+			if source == ConfigSourceRemote && !a.config.EnableScriptChecks && !a.config.EnableRemoteScriptChecks {
```
Comment (pierresouchay) [inline agent/agent.go]: Same

**Comment by pierresouchay** [comment]
@Aestek There is a weird compilation error: https://travis-ci.org/hashicorp/consul/jobs/433508308

Unfortunately Travis is using golang 1.11.x (see https://github.com/hashicorp/consul/pull/4496) which is less permissive than Golan 1.10 - might be linked

**Comment by Aestek** [comment]
The build error is weird indeed. The one job that failed ran with go 1.7. `math/bits` was added in go 1.9.

File: agent/agent.go (hunk 1 above)
Comment (Aestek) [reply agent/agent.go]: :+1:

File: agent/agent.go (hunk 2 above)
Comment (Aestek) [reply agent/agent.go]: :+1:

**Comment by pierresouchay** [comment]
@pearkes I think this change is both simple and important for clusters where you need script checks but don't want to open your system to arbitrary remote execution vulnerability.

We recently got bitten by this and I suspect it would be a great addition in many many shops.

**Comment by pierresouchay** [comment]
@mkeeler Your change seems reasonnable. @Aestek Do you think you can implement this?

**Comment by Aestek** [comment]
@mkeeler Thanks for looking at this. I've updated the PR to remove `enable_remote_script_checks`, leaving only `enable_script_checks` and `enable_local_script_checks` as you suggested.

**Comment by mkeeler** [comment]
@Aestek can you rebase this to master. There is a conflict in agent/agent.go and one other change in agent/agent_endpoint.go that needs to be made for where AddService is called.

**Comment by Aestek** [comment]
@mkeeler done

**Comment by banks** [comment]
I just fixed one more whitespace conflict in docs following last merge. Will wait for CI to be green but this should be good to merge.

**Pull request merged**